File: src/main/kotlin/net/merayen/elastic/backend/architectures/local/nodes/group_1/LProcessor.java
package net.merayen.elastic.backend.architectures.local.nodes.group_1;

import net.merayen.elastic.backend.architectures.local.LocalProcessor;
import net.merayen.elastic.backend.architectures.local.exceptions.SpawnLimitException;

public class LProcessor extends LocalProcessor {
	int session_id = -1;

	@Override
	protected void onInit() {}

	@Override
	protected void onPrepare() {}

	@Override
	protected void onProcess() {
		
			try {
				if(session_id == -1)
					session_id = spawnSession(0);
			} catch (SpawnLimitException e) {
				throw new RuntimeException("Should not happen");
			}
	}

	@Override
	protected void onMessage(Object message) {}

	@Override
	protected void onDestroy() {}

}

File: src/main/kotlin/net/merayen/elastic/backend/architectures/local/nodes/midi_1/LProcessor.kt
package net.merayen.elastic.backend.architectures.local.nodes.midi_1

import net.merayen.elastic.backend.architectures.local.GroupLNode
import net.merayen.elastic.backend.architectures.local.LocalProcessor
import net.merayen.elastic.backend.architectures.local.lets.MidiOutlet
import net.merayen.elastic.system.intercom.ElasticMessage

class LProcessor : LocalProcessor() {
	override fun onInit() {}

	override fun onPrepare() {}

	override fun onProcess() {
		val outlet = getOutlet("out") as MidiOutlet?
		if (outlet != null) {
			val parent = localNode.parent as GroupLNode

			val startBeat = parent.getCursorPosition()

			val stopBeat = if (parent.isPlaying()) {
				startBeat + (parent.getCurrentFrameBPM() * (buffer_size / sample_rate.toDouble())) / 60.0
			} else {
				startBeat
			}

			if (parent.isPlaying())
				println("Playing! startBeat=${startBeat}, stopBeat=${stopBeat}")

			val lnode = (localNode as LNode)
			val inputMidi = lnode.inputMidi
			val midiData = lnode.midiData

			if (inputMidi != null) // TODO write any note from MidiData too
				outlet.putMidi(0, inputMidi); // TODO shouldn't quantize to beginning of frame

			outlet.written = buffer_size
			outlet.push()
		}
	}

	override fun onMessage(message: ElasticMessage) {}

	override fun onDestroy() {}
}
